src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/SystemClassWithMembersAndTypesRecord.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.Runtime.Serialization.BinaryFormat.Utils;
+
+namespace System.Runtime.Serialization.BinaryFormat;
+
+internal sealed class SystemClassWithMembersAndTypesRecord : ClassRecord
+{
+    private SystemClassWithMembersAndTypesRecord(ClassInfo classInfo, MemberTypeInfo memberTypeInfo)
+        : base(classInfo, memberTypeInfo)
+    {
+    }
+
+    public override RecordType RecordType => RecordType.SystemClassWithMembersAndTypes;
+
+    public override bool IsTypeNameMatching(Type type)
+        => type.Assembly == typeof(object).Assembly
+        && type.GetTypeFullNameIncludingTypeForwards() == ClassInfo.TypeName.FullName;
+
+    internal static SystemClassWithMembersAndTypesRecord Parse(BinaryReader reader, RecordMap recordMap, PayloadOptions options)
+    {
+        ClassInfo classInfo = ClassInfo.Parse(reader);
+        MemberTypeInfo memberTypeInfo = MemberTypeInfo.Parse(reader, classInfo.MemberNames.Count, options, recordMap);
+        // the only difference with ClassWithMembersAndTypesRecord is that we don't read library id here
+        classInfo.ParseTypeName(options);
+        return new(classInfo, memberTypeInfo);
+    }
+
+    internal override (AllowedRecordTypes allowed, PrimitiveType primitiveType) GetNextAllowedRecordType()
+        => MemberTypeInfo.GetNextAllowedRecordType(MemberValues.Count);
+
+    // For the root records that turn out to be primitive types, we map them to
+    // PrimitiveTypeRecord<T> so the users don't need to learn the BF internals
+    // to get a single primitive value!
+    internal SerializationRecord TryToMapToUserFriendly()
+    {
+        // It could be implemented with way fewer ifs, but perf is important
+        // and we want to bail out as soon as possible, but also don't convert something
+        // that is not an exact match.
```

Comment:
> but perf is important

Not particularly :smile:.  Sorting by the length of MemberValues makes sense, and helps with the correctness aspect of "exact match".  The perf argument isn't something that needs to live in the code.

I'd delete the whole comment, honestly.